server.go

```diff
@@ -223,6 +279,14 @@ func (server *mongoServer) RecycleSocket(socket *mongoSocket) {
 	if !server.closed {
 		server.unusedSockets = append(server.unusedSockets, socket)
 	}
+	// If anybody is waiting for a connection, they should try now.
+	// Note that this _has_ to be broadcast, not signal; the signature of AcquireSocket
+	// and AcquireSocketWithBlocking allow the caller to specify the max number of connections,
+	// rather than that being an intrinsic property of the connection pool (I assume to ensure
+	// that there is always a connection available for replset topology discovery). Thus, once
+	// a connection is returned to the pool, _every_ waiter needs to check if the connection count
+	// is underneath their particular value for poolLimit.
+	server.poolWaiter.Broadcast()
```

Comment:
Again, I am not 100% sure of my reading of the code ( mgo was never super fun to read, and I haven't had a closer look at it since some time ago, but you take a lock on the server when you try to acquire a socket, and block wait on a signal. The signal won't be raised, if you first need to take a lock on the same mutex, no ?

Before, the pool was unlocked before returning an ErrPoolLimit, but it is not the case anymore.

Reply:
> mgo was never super fun to read

Yyyyyyup......

> The signal won't be raised, if you first need to take a lock on the same mutex, no ?

I _think_ my answer above covers this - the acquiring code implicitly releases the mutex in the call to `Wait` and re-acquires it when `Wait` returns. So the release code should be able to grab the mutex and send the signal.

>  Before, the pool was unlocked before returning an ErrPoolLimit, but it is not the case anymore.

I think i'm calling `Unlock()` on line 167:

```
} else {
	if len(server.liveSockets)-len(server.unusedSockets) >= poolLimit {
		server.Unlock()
		return nil, false, errPoolLimit
	}
}
```

Is this not sufficient?

Reply:
Yes, it is. I'd like to blame the low caffeine level in my blood for raising this issue ;)